core/src/main/scala/kafka/message/Message.scala

```diff
@@ -36,83 +39,156 @@ object Message {
   val MagicLength = 1
   val AttributesOffset = MagicOffset + MagicLength
   val AttributesLength = 1
-  val KeySizeOffset = AttributesOffset + AttributesLength
+  // Only message format version 1 has the timestamp field.
+  val TimestampOffset = AttributesOffset + AttributesLength
+  val TimestampLength = 8
+  // Message format version 0 does not have timestamp field.
+  val KeySizeOffset_V0 = AttributesOffset + AttributesLength
+  // Message format version 1 has timestamp field.
+  val KeySizeOffset_V1 = TimestampOffset + TimestampLength
   val KeySizeLength = 4
-  val KeyOffset = KeySizeOffset + KeySizeLength
+  // Message format version 0 does not have timestamp field.
+  val KeyOffset_V0 = KeySizeOffset_V0 + KeySizeLength
+  // Message format version 1 has timestamp field.
+  val KeyOffset_V1 = KeySizeOffset_V1 + KeySizeLength
   val ValueSizeLength = 4
-  
-  /** The amount of overhead bytes in a message */
-  val MessageOverhead = KeyOffset + ValueSizeLength
+
+  private val MessageHeaderSizeMap = Map (
+    0.asInstanceOf[Byte] -> (CrcLength + MagicLength + AttributesLength + KeySizeLength + ValueSizeLength),
+    1.asInstanceOf[Byte] -> (CrcLength + MagicLength + AttributesLength + TimestampLength + KeySizeLength + ValueSizeLength))
+
+  /**
+   * The amount of overhead bytes in a message
+   * This value is only used to check if the message size is valid or not. So the minimum possible message bytes is
+   * used here, which comes from a message in message format V0 with empty key and value.
+   */
+  val MinMessageOverhead = KeyOffset_V0 + ValueSizeLength
   
   /**
    * The minimum valid size for the message header
+   * The MinHeaderSize does not include the TimestampLength for backward compatibility.
+   * This value is only used to check if the message size is valid or not. So the minimum possible message bytes is
+   * used here, which comes from message format V0.
    */
-  val MinHeaderSize = CrcLength + MagicLength + AttributesLength + KeySizeLength + ValueSizeLength
+  val MinHeaderSize = MessageHeaderSizeMap.values.min
```

Comment:
It seems that MinHeaderSize is the same as MinMessageOverhead. Could we consolidate them?